api/envoy/config/rbac/v2alpha/rbac.proto

```diff
@@ -0,0 +1,87 @@
+syntax = "proto3";
+
+import "validate/validate.proto";
+
+package envoy.config.rbac.v2alpha;
+
+// Role Based Access Control (RBAC) provides service-level and method-level access control for a service.
+// Here is an example of RBAC configuration.
+//
+//   action_type: ALLOW
+//   policies:
+//     "service-admin":
+//       permissions:
+//       - methods: [simple:"*"]
+//       subjects:
+//       - name: "cluster.local/ns/default/sa/admin"
+//     "product-viewer":
+//       permissions:
+//       - paths: [prefix:"/products", suffix:"/reviews"]
+//         methods: ["GET"]
+//       subjects:
+//       - name: "cluster.local/ns/default/sa/product-client"
+//
+message RBAC {
+  // Should we do white-list or black-list style access control.
+  enum ActionType {
+    // The policies grant access to subjects. The rest is denied. This is white-list style
+    // access control. This is the default type.
+    ALLOW = 0;
+
+    // The policies deny access to subjects. The rest is allowed. This is black-list style
+    // access control.
+    DENY = 1;
+  }
+
+  ActionType action_type = 1;
+
+  // Maps from policy name to policy.
+  map<string, Policy> policies = 2;
+}
+
+// Policy specifies a role and the subjects that are assigned/denied the role.
+message Policy {
+ // Required. The set of permissions that define a role.
+ repeated Permission permissions = 1 [(validate.rules).repeated .min_items = 1];
+
+ // Required. List of subjects that are assigned/denied the role based on “action_type”.
+ repeated Subject subjects = 2 [(validate.rules).repeated .min_items = 1];
```

Comment:
Per our offline discussions, a few comments:
1) WDYT about s/Subject/Principal? I think Principal is the more generally accepted term and this is what AttributeContext already uses?

2) I think we agreed that we are going to make Principal a first class thing in Envoy, such that every connection has an attached Principal. Other filters can set the Principal. However, I think we need to have first class support for the "unknown" Principal. I think there are a few ways we can do that, but what are your thoughts here? On option is to just allow this list to be empty. If there are no principals the policy applies to all/unknown principals?

Reply:
+1 to use principal, which is a well understood term to identify the caller.

I wonder if we still need a high-level message object that groups principal fields and other attributes/metadata to identify the caller. "subject" term is borrowed from Kubernetes RBAC. Note sure if we can come up with a better term to group principal and other attributes/metadata

Reply:
@mattklein123, or are you suggesting replacing the Subject with Principal as the high-level message object where we can put name and other attributes? I am fine with that change in this case

Reply:
Yeah I'm literally suggesting s/Subject/Principal everywhere...

Reply:
+1

Reply:
1. I renamed "Subject" to "Principal". 

2. I added comment that if "principal.name" is set to "*", it applies to any user. This addresses the unknown/all principal case you mentioned. This is also the way Istio RBAC does today. I showed an example of assigning "product-viewer" role to "any user" in the beginning of rbac.proto file.